### PR TITLE
`.formula_list_to_named_list()` improvements IN PROGRESS

### PR DESCRIPTION
**Pull request by ddsjoberg**
- Added checks to `.formula_list_to_named_list(x=)` type to ensure a list or formula has been passed. Check triggers improved messaging instructing users of the correct syntax.
- Added a `.formula_list_to_named_list(type_check=)` argument. Argument accepts a predicate function as the input and checks the RHS of the formula passes the predicate check.

### CODE REVIEW

**Comment by larmarange** [comment]
Quick question, should we have a `type_check_expected_message` argument allowing to pass some explanations displayed in the error message saying what is expected?

File: tests/testthat/test-select_helpers.R [hunk 1]
```diff
@@ -291,6 +305,23 @@ test_that("select_helpers: .formula_list_to_named_list ", {
...
   )
+
+  expect_error(
+    .formula_list_to_named_list(list(age ~ "Age"), var_info = tidy_mod, type_check = is.logical,
+                                arg_name = "label")
+  )
+  expect_error(
+    .formula_list_to_named_list(list(age ~ "Age"), var_info = tidy_mod,
+                                type_check = is.logical, arg_name = "label",
+                                type_check_msg = "Age msg error")
```
Comment (larmarange) [inline tests/testthat/test-select_helpers.R:316]: we should test if the error message contains "Age msg error"?

File: tests/testthat/test-select_helpers.R [hunk 2]
```diff
@@ -291,6 +305,23 @@ test_that("select_helpers: .formula_list_to_named_list ", {
...
+  )
+  expect_error(
+    .formula_list_to_named_list("Age", var_info = tidy_mod,
+                                type_check = rlang::is_string,
+                                arg_name = "label",
+                                type_check_msg = "Age msg error")
+  )
```
Comment (larmarange) [inline tests/testthat/test-select_helpers.R:323]: Ibid

File: tests/testthat/test-select_helpers.R (hunk 1 above)
Comment (ddsjoberg) [reply tests/testthat/test-select_helpers.R:316]: good idea :)

File: tests/testthat/test-select_helpers.R (hunk 2 above)
Comment (ddsjoberg) [reply tests/testthat/test-select_helpers.R:323]: image

This is the error message, but I was having trouble getting a regex match for the message. Do you know why this doesn't work?

```r
d>   expect_error(
+     .formula_list_to_named_list("Age", var_info = tidy_mod,
+                                 type_check = rlang::is_string,
+                                 arg_name = "label",
+                                 type_check_msg = "Age msg error"),
+     "*Did you mean `everything() ~ \"Age\"`\\?$"
+   )
Error: `.formula_list_to_named_list(...)` threw an error with unexpected message.
Expected match: "*Did you mean `everything() ~ \"Age\"`\\?$"
Actual message: "Error processing the `label` argument. Expecting a list or formula.\nReview syntax details at'https://www.danieldsjoberg.com/gtsummary/reference/syntax.html'\n\nDid you mean `everything() ~ \"Age\"`?"
```

Comment (larmarange) [reply tests/testthat/test-select_helpers.R:325]: This is normal because it was not a formula.

**Review by larmarange — COMMENTED** [review]
We just need to update NEWS.

File: tests/testthat/test-select_helpers.R (hunk 2 above)
Comment (ddsjoberg) [reply tests/testthat/test-select_helpers.R:329]: This is testing the error message as well. Here the logic is testing that the input is indeed a list (or a formula). But in this example the user passed the string `"Age"`. I added logic that if the users passes an incorrect type (ie not a list or a formula) AND the type of the passed object matched the `type_check=` predicate, the message asks the user 'Did you mean `everything() ~ "Age"`?'

**Comment by larmarange** [comment]
OK I just updated the test
-- 
*Joseph LARMARANGE*
Demographer PhD at IRD <http://www.ird.fr/> · Ceped <http://www.ceped.org/> UMR
196 (University of Paris - IRD) · SageSud ERL Inserm 1244



Tel. (Ceped): +33 1 76 53 34 75
Cell. phone (France) : +33 6 62 06 51 82
Cell. Phone (Côte d'Ivoire): +225 07 89 40 59 33
Skype: joseph.larmarange
Ceped, 45 rue des Saints-Pères, 75006 Paris, France


Le lun. 6 déc. 2021 à 14:37, Daniel Sjoberg ***@***.***> a
écrit :

**Pull request merged**